src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt

```diff
@@ -21,38 +25,48 @@ import uk.gov.justice.digital.hmpps.externalusersapi.security.MaintainUserCheck
 @Transactional(readOnly = true)
 class GroupsService(
   private val groupRepository: GroupRepository,
-  private val maintainUserCheck: MaintainUserCheck,
   private val childGroupRepository: ChildGroupRepository,
   private val telemetryClient: TelemetryClient,
   private val authenticationFacade: AuthenticationFacade,
   private val userRepository: UserRepository,
   private val userGroupService: UserGroupService,
+  private val groupAssignableRoleRepository: GroupAssignableRoleRepository,
+  private val maintainUserCheck: MaintainUserCheck,
 ) {
+  // TODO Check: can we amend this call to work
 
   val allGroups: List<Group>
-    get() = groupRepository.findAllByOrderByGroupName()
+    get() = runBlocking { groupRepository.findAllByOrderByGroupName().toList() }
+
+  suspend fun getAllGroups(): List<Group> =
+    coroutineScope {
+      val groups = async {
+        groupRepository.findAllByOrderByGroupName()
+      }
+      groups.await().toList()
+    }
 
   @Throws(GroupNotFoundException::class)
-  fun getGroupDetail(groupCode: String): Group {
-    val requestedGroup =
-      groupRepository.findByGroupCode(groupCode) ?: throw GroupNotFoundException("get", groupCode, "notfound")
-
-    Hibernate.initialize(requestedGroup.assignableRoles)
-    Hibernate.initialize(requestedGroup.children)
-    maintainUserCheck.ensureMaintainerGroupRelationship(authenticationFacade.currentUsername, groupCode)
-    return requestedGroup
-  }
+  suspend fun getGroupDetail(groupCode: String): GroupDetails =
+    groupRepository.findByGroupCode(groupCode)
+      ?.let {
+        val children = childGroupRepository.findAllByGroup(it.groupId).toList()
+        val assignableRole = groupAssignableRoleRepository.findGroupAssignableRoleByGroupCode(it.groupCode).toList()
+        maintainUserCheck.ensureMaintainerGroupRelationship(authenticationFacade.getUsername(), groupCode)
```

Comment:
GroupsService calls back to UserService via MaintainUserCheck - feels bit convoluted? Could this code be re-structured at all to improve this / reduce the number of queries?

Reply:
I agree

Reply:
TODO 14 -  See if possible to rewrite/improve the code